Clerk Elements (Beta) - Initial docs (#890)
clerk/clerk-docs: docs/elements/guides/sign-up.mdx

---
description: Learn how to build a complete sign-up form with Clerk Elements.
---

> # Build a sign-up flow

Reviewer: All my suggestions for this page are the same as my suggestions for Please use the suggestion change I made in [the sign-in guide](https://github.com/clerk/clerk-docs/pull/890/files/3bd5d665cc7641c0d2db48655b79b84bbe48cec0#diff-9addefb83500d3a49204d58b6636b2020b87608132d4aec8e1451e79cacefa25R2) above.

Once we've concluded discussions for those changes, I can pull this PR and add them to this file, saving us all time 👍

Reviewer 2: Applied your suggestions to both guides 👍

> In this guide you'll learn how to build a complete sign-up flow using Clerk Elements components.

Reviewer 3: ```suggestion
In this guide you will learn how to build a complete sign-up flow using Clerk Elements components.
```

> In this guide you will learn how to build a complete sign-up flow using Clerk Elements components.

Reviewer: ```suggestion

<TutorialHero
  beforeYouStart={[
    {
      title: "Set up a Clerk application",
      link: "https://clerk.com/docs/quickstarts/setup-clerk",
    },
    {
      title: "Create a Next.js application",
      link: "https://nextjs.org/docs/getting-started/installation",
    },
    {
      title: "Upgrade to Core 2 if necessary",
      link: "https://clerk.com/docs/upgrade-guides/core-2/nextjs#upgrading-clerk-nextjs-to-core-2",
    },
  ]}
>
- Build a complete sign-up flow using Clerk Elements
</TutorialHero>

```

<Callout type="info">

> Clerk Elements currently only works with Next.js and [Clerk Core 2](https://clerk.com/changelog/2024-04-19). As it gets closer to a stable release, support for additional frameworks will be added. If your Next.js application is already using Clerk, make sure to [upgrade to Core 2](/docs/upgrade-guides/core-2/nextjs). If you're starting from scratch, follow the [Next.js quickstart](/docs/quickstarts/nextjs) before proceeding.

Reviewer 3: ```suggestion
Clerk Elements currently only works with Next.js and [Clerk Core 2](https://clerk.com/changelog/2024-04-19). As it gets closer to a stable release, support for additional frameworks will be added. If your Next.js application is already using Clerk, make sure to [upgrade to Core 2](/docs/upgrade-guides/core-2/nextjs). If you are starting from scratch, follow the [Next.js quickstart](/docs/quickstarts/nextjs) before proceeding.
```

> Clerk Elements currently only works with Next.js and [Clerk Core 2](https://clerk.com/changelog/2024-04-19). As it gets closer to a stable release, support for additional frameworks will be added. If your Next.js application is already using Clerk, make sure to [upgrade to Core 2](/docs/upgrade-guides/core-2/nextjs). If you are starting from scratch, follow the [Next.js quickstart](/docs/quickstarts/nextjs) before proceeding.

Reviewer: Please use the suggestion change I made for this in [the sign-in guide](https://github.com/clerk/clerk-docs/pull/890/files/3bd5d665cc7641c0d2db48655b79b84bbe48cec0#diff-9addefb83500d3a49204d58b6636b2020b87608132d4aec8e1451e79cacefa25R2) above

</Callout>

<Steps>

### Add a sign-up route

Create a new route in your Next.js application. The route needs to be an [optional catch-all route](https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes#optional-catch-all-segments) so the sign-up flow can handled nested paths. Add the following snippet to `/app/sign-up/[[...sign-up]]/page.tsx`:

```tsx filename="/app/sign-up/[[...sign-up]]/page.tsx"
'use client'

import * as Clerk from '@clerk/elements/common'
import * as SignUp from '@clerk/elements/sign-up'

export default function SignUpPage() {
  return (
    <SignUp.Root>
    </SignUp.Root>
  )
}
```

> You've also added two imports, which you'll use as you build out the rest of the flow. `<SignUp.Root>` manages the sign-up state and handles connecting the components to Clerk's APIs.

Reviewer 3: ```suggestion
You will use these two imports to build out the rest of the flow. `<SignUp.Root>` manages the sign-up state and handles connecting the components to Clerk's APIs.
```

### Add the start step

The Clerk authentication flows are made up of **steps**. Steps handle rendering the UI for each part of the flow. To allow users to create a sign-up attempt, the `start` step needs to be rendered. Use the `<SignUp.Step>` component to render this:

```tsx filename="/app/sign-up/[[...sign-up]]/page.tsx" {9-11}
'use client'

import * as Clerk from '@clerk/elements/common'
import * as SignUp from '@clerk/elements/sign-up'

export default function SignUpPage() {
  return (
    <SignUp.Root>
      <SignUp.Step name="start">
        <h1>Create an account</h1>
      </SignUp.Step>
    </SignUp.Root>
  )
}
```

### Add form fields

> Make it functional by adding fields. For the sake of this guide, let's assume that your instance is configured to require a username and allow authentication with email codes, as well as with Google. You'll use the Clerk Elements field components to render `identifier` and `username` fields, as well as the `<Connection>` component to render a button that allows users to sign up with Google.

Reviewer 3: ```suggestion
Make it functional by adding fields. For the sake of this guide, let's assume that your instance is configured to require a username and allow authentication with email codes, as well as with Google. You will use the Clerk Elements field components to render `identifier` and `username` fields, as well as the `<Connection>` component to render a button that allows users to sign up with Google.
```

import * as Clerk from '@clerk/elements/common'
import * as SignUp from '@clerk/elements/sign-up'

export default function SignUpPage() {
  return (
    <SignUp.Root>
      <SignUp.Step name="start">
        <h1>Create an account</h1>

        <Clerk.Connection name="google">
          Sign up with Google
        </Clerk.Connection>

        <Clerk.Field name="identifier">
          <Clerk.Label>Email</Clerk.Label>
          <Clerk.Input />
          <Clerk.FieldError />
        </Clerk.Field>

        <Clerk.Field name="username">
          <Clerk.Label>Username</Clerk.Label>
          <Clerk.Input />
          <Clerk.FieldError />
        </Clerk.Field>

        <SignUp.Action submit>Sign up</SignUp.Action>
      </SignUp.Step>
    </SignUp.Root>
  )
}
```

`<Clerk.Field>` takes care of wiring up the input with the label element, and `<Clerk.FieldError>` will render any field-specific errors that get returned from the Clerk's API. You've also used the `<SignUp.Action>` component. This component provides common actions that are used throughout the flows. In this case, you're using the `submit` action to render a submit button for the start form.

### Accept additional fields

> If a user now attempts to sign up via Google, but username is a required field, you'll need to render the `continue` step to accept the username. Use the `<SignUp.Step>` component to render additional required fields.

Reviewer 3: ```suggestion
Should a user attempt to sign up via Google while a username is a required field, the ﻿`continue` step will be necessary to accept the username. The ﻿`<SignUp.Step>` component will display any additional required fields.
```

        <SignUp.Action submit>Sign up</SignUp.Action>
      </SignUp.Step>
      <SignUp.Step name="continue">
        <h1>Fill in missing fields</h1>

        <Clerk.Field name="username">
          <Clerk.Label>Username</Clerk.Label>
          <Clerk.Input />
          <Clerk.FieldError />
        </Clerk.Field>

        <SignUp.Action submit>Continue</SignUp.Action>
      </SignUp.Step>
      <SignUp.Step name="verifications">
        <SignUp.Strategy name="phone_code">
          <h1>Check your phone for an SMS</h1>

          <Clerk.Field name="code">
            <Clerk.Label>Phone Code</Clerk.Label>
            <Clerk.Input />
            <Clerk.FieldError />
          </Clerk.Field>

          <SignUp.Action submit>Verify</SignUp.Action>
        </SignUp.Strategy>
        <SignUp.Strategy name="email_code">
          <h1>Check your email</h1>

          <Clerk.Field name="code">
            <Clerk.Label>Email Code</Clerk.Label>
            <Clerk.Input />
            <Clerk.FieldError />
          </Clerk.Field>

          <SignUp.Action submit>Verify</SignUp.Action>
        </SignUp.Strategy>
      </SignUp.Step>
    </SignUp.Root>
  )
}
```

### Customize and add styling

> You might have noticed by now that none of components rendered so far come with their own styling. This is by design! With the exception of `<SignUp.Root>` and `<SignUp.Strategy>`, every component rendered so far can be styled. You can use the `className` prop to add custom classes to each component. Check out the [styling guide](/docs/elements/guides/styling) for a more in-depth guide on styling Clerk Elements.

Reviewer 3: ```suggestion
You may have noticed that none of the components displayed to this point include styling. This is intentional! With the exception of `<SignUp.Root>` and `<SignUp.Strategy>`, every component rendered so far can be styled. The ﻿`className` prop enables the addition of custom classes to each component. Check out the [styling guide](/docs/elements/guides/styling) for a more in-depth guide on styling Clerk Elements.
```

> If you're looking to customize the UI even further, check out additional Clerk Elements components like [`<Loading>`](/docs/elements/reference/common#loading) and [`<FieldState>`](/docs/elements/reference/common#fieldstate).

Reviewer 3: ```suggestion
For more extensive customization of the UI, check out additional Clerk Elements components such as [`<Loading>`](/docs/elements/reference/common#loading) and [`<FieldState>`](/docs/elements/reference/common#fieldstate).
```
